src/backends/android.rs

```diff
@@ -0,0 +1,160 @@
+    pub fn resize(&mut self, width: NonZeroU32, height: NonZeroU32) -> Result<(), SoftBufferError> {
+        let (width, height) = (|| {
+            let width = NonZeroI32::try_from(width).ok()?;
+            let height = NonZeroI32::try_from(height).ok()?;
+            Some((width, height))
+        })()
+        .ok_or(SoftBufferError::SizeOutOfRange { width, height })?;
+
+        // Do not change the format.
+        self.native_window
+            .set_buffers_geometry(
+                width.into(),
+                height.into(),
+                // Default is typically R5G6B5 16bpp, switch to 32bpp
+                Some(HardwareBufferFormat::R8G8B8A8_UNORM),
+            )
+            .map_err(|err| {
+                SoftBufferError::PlatformError(
+                    Some("Failed to set buffer geometry on ANativeWindow".to_owned()),
+                    Some(Box::new(err)),
+                )
+            })
+    }
+
+    pub fn buffer_mut(&mut self) -> Result<BufferImpl<'_, D, W>, SoftBufferError> {
+        let lock_guard = self.native_window.lock(None).map_err(|err| {
+            SoftBufferError::PlatformError(
+                Some("Failed to lock ANativeWindow".to_owned()),
+                Some(Box::new(err)),
+            )
+        })?;
+
+        assert_eq!(
+            lock_guard.format().bytes_per_pixel(),
+            Some(4),
+            "Unexpected buffer format {:?}, please call .resize() first to change it to RGBA8888",
+            lock_guard.format()
+        );
+
+        Ok(BufferImpl(lock_guard, PhantomData, PhantomData))
+    }
+
+    /// Fetch the buffer from the window.
+    pub fn fetch(&mut self) -> Result<Vec<u32>, SoftBufferError> {
+        Err(SoftBufferError::Unimplemented)
+    }
+}
+
+pub struct BufferImpl<'a, D: ?Sized, W>(
+    NativeWindowBufferLockGuard<'a>,
+    PhantomData<&'a D>,
+    PhantomData<&'a W>,
+);
+
+// TODO: Move to NativeWindowBufferLockGuard?
+unsafe impl<'a, D, W> Send for BufferImpl<'a, D, W> {}
+
+impl<'a, D: HasDisplayHandle + ?Sized, W: HasWindowHandle> BufferImpl<'a, D, W> {
+    #[inline]
+    pub fn pixels(&self) -> &[u32] {
+        todo!()
+        // unsafe {
+        //     std::slice::from_raw_parts(
+        //         self.0.bits().cast_const().cast(),
+        //         (self.0.stride() * self.0.height()) as usize,
+        //     )
+        // }
+    }
+
+    #[inline]
+    pub fn pixels_mut(&mut self) -> &mut [u32] {
+        let bytes = self.0.bytes().expect("Nonplanar format");
+        unsafe {
+            std::slice::from_raw_parts_mut(
+                bytes.as_mut_ptr().cast(),
+                bytes.len() / std::mem::size_of::<u32>(),
+            )
+        }
+    }
+
+    /// The number of _pixels_ that a line in the buffer takes in memory.
+    #[inline]
+    pub fn stride(&self) -> u32 {
+        self.0.stride() as u32
+    }
+
+    pub fn age(&self) -> u8 {
+        todo!()
```

Comment:
Android works with buffer pools and buffer queues, to implement multi-buffering. The [underlying function](https://developer.android.com/ndk/reference/group/a-native-window#group___a_native_window_1ga0b0e3b7d442dee83e1a1b42e5b0caee6) is defined as:

> Lock the window's next drawing surface for writing.

Previous data is probably undefined, if a different handle is returned most of the time. However, I've implemented a `Vec` with a copy of the data, which we might use for this (and to implement the new `fn fetch()`?).

---

In a quick test I just kept track of the raw pointers returned from this function. After 3 calls the pointer repeats, and always in the same order. If I don't write to this pointer the old data is preserved.

I have no clue if this is something we can rely on however, since Android doesn't specify the lifetime of this CPU handle. And depending on the phone hardware, display presentation unit and UMA, may or may not have to copy this CPU visible memory (and discard it right after!) to make it presentable.

Allow me to go on a tangent and quickly check `dumpsys SurfaceFlinger`:

```
---------------------------------------------------------------------------------------------------------------------------------------------------------------
 rust.example.winit_android/android.app.NativeActivity#39518
  rel      0 |            1 |            0 |     DEVICE |          0 |    0    0 1096 2560 |    0.0    0.0 1096.0 2560.0 |                                              [*]
...
HWC2 display_id: 0
layer: 11369 name:                                                               VRI[NativeActivity]#0(BLAST Consumer)0 z: 0 composition: Device/Device alpha: 255 format:              RGBX_8888 dataspace:0x08810000 transform: 0/0/0 buffer_id: 0xb400007871ca9ab0 secure: 0
                 LayerBuffer:  width=[1152],  height=[2560],  unaligned_width=[1096],  unaligned_height=[2560]
                 LayerBufferFlags:  secure=[0],  interlace=[0],  hdr=[0]
                 ColorMetaData:  colorPrimaries=[1],  range=[1],  transfer=[1],  matrixCoefficients=[0](matrixCoefficients is not used for SDR layer)
...
|-----|---------------|-----------|------|-------------|--------------------------|---------------------|---------------------|----|------------|-----------|----|-----|----|
| Idx |   Comp Type   |   Split   | Pipe |    W x H    |          Format          |  Src Rect (L T R B) |  Dst Rect (L T R B) |  Z | Pipe Flags | Deci(HxV) | CS | Rng | Tr |
|-----|---------------|-----------|------|-------------|--------------------------|---------------------|---------------------|----|------------|-----------|----|-----|----|
|   0 |           SDE |    Pipe-1 |  139 | 1152 x 2560 |                RGBX_8888 |    0    0  548 2560 |    0    0  548 2560 |  0 | 0x00002001 |   0 x   0 |  1 |   1 |  1 |
|     |               |    Pipe-2 |  136 | 1152 x 2560 |                RGBX_8888 |  548    0 1096 2560 |  548    0 1096 2560 |  0 | 0x00002001 |   0 x   0 |  1 |   1 |  1 |
```

If you ignore the fact that my phone is a dualpipe (two layer mixers and two encoders, each getting half of the plane, to drive a 4k@120Hz screen), you can clearly see that the "composition type" of this `RGBX_8888` buffer of `1152x2560` pixels is "Snapdragon Display Engine" instead of some kind of software (or GPU) composition.

---

Completely separate from all this there's a more low-level API to [dispatch `HardwareBuffer`s (that we allocate ourselves!) to a surface](https://developer.android.com/ndk/reference/group/native-activity#asurfacetransaction_setbuffer) [^1]. Here we _could_ implement buffer aging correctly.

[^1]: Lots of caveats apply here:
      1. The API is still on a branch for the `ndk` crate, to be finished;
      2. NDK functions to retrieve a `SurfaceControl` for the root `NativeWindow` are broken right now and might only be fixed in Android 15 onwards;
      3. Requires API level 29.